numba/cuda/cudadrv/driver.py

```diff
@@ -775,8 +775,7 @@ def _module_finalizer(context, handle):
 
     def core():
         def cleanup():
-            if modules:
-                del modules[handle.value]
+            assert handle.value not in modules
```

Comment:
> Can you add a comment explaining why this is the case? 

Added comments in https://github.com/numba/numba/pull/1859/commits/c5ac9b10664dbcd433479f9e3cf3e7adae92833b

> Also, why is the call to cuModuleUnload() made here rather than in _module_finalizer() itself?

All CUDA resource deallocation in this file is detached from the object lifetime and handled in the "trashing service" to avoid operating on a corrupted CUDA context which may result in segfault or worse (crashing the OSX kernel as I have encountered).